pow/src/test/java/tech/pegasys/artemis/pow/DepositRequestManagerTest.java

```diff
@@ -0,0 +1,266 @@
+    verify(depositEventChannel).notifyDepositsFromBlock(argThat(isEvent(4, 1)));
+    verify(depositEventChannel).notifyDepositsFromBlock(argThat(isEvent(10, 1)));
+    verify(depositEventChannel).notifyDepositsFromBlock(argThat(isEvent(11, 1)));
+    verify(depositEventChannel).notifyDepositsFromBlock(argThat(isEvent(14, 1)));
+    verifyNoMoreInteractions(depositEventChannel);
+  }
+
+  @Test
+  void runSecondAttemptWhenFirstAttemptFails() {
+
+    SafeFuture<List<DepositContract.DepositEventEventResponse>> firstRequestFuture =
+        mockContractEventsInRange(0, 10);
+
+    mockBlockForEth1Provider("0x0001", 1, 1000);
+    mockBlockForEth1Provider("0x0002", 2, 1014);
+
+    depositRequestManager.start();
+
+    pushLatestCanonicalBlockWithNumber(10);
+
+    firstRequestFuture.completeExceptionally(new RuntimeException("Nope"));
+
+    System.out.println(asyncRunner.hasDelayedActions());
+
+    SafeFuture<List<DepositContract.DepositEventEventResponse>> secondRequestFuture =
+        mockContractEventsInRange(0, 10);
+
+    secondRequestFuture.complete(
+        List.of(
+            mockDepositEventEventResponse(1, "0x0001", 1),
+            mockDepositEventEventResponse(2, "0x0002", 2)));
+
+    asyncRunner.executeQueuedActions();
+
+    verify(depositEventChannel).notifyDepositsFromBlock(argThat(isEvent(1, 1)));
+    verify(depositEventChannel).notifyDepositsFromBlock(argThat(isEvent(2, 1)));
+    verifyNoMoreInteractions(depositEventChannel);
+  }
+
+  private void mockBlockForEth1Provider(String blockHash, long blockNumber, long timestamp) {
+    EthBlock.Block block = mock(EthBlock.Block.class);
+    when(block.getTimestamp()).thenReturn(BigInteger.valueOf(timestamp));
+    when(block.getNumber()).thenReturn(BigInteger.valueOf(blockNumber));
+    when(block.getHash()).thenReturn(blockHash);
+    when(eth1Provider.getEth1BlockFuture(blockHash)).thenReturn(SafeFuture.completedFuture(block));
+  }
+
+  private void pushLatestCanonicalBlockWithNumber(long latestBlockNumber) {
+    EthBlock.Block block = mock(EthBlock.Block.class);
+    when(block.getNumber())
+        .thenReturn(
+            BigInteger.valueOf(latestBlockNumber).add(ETH1_FOLLOW_DISTANCE.bigIntegerValue()));
+    blockPublisher.onNext(block);
+  }
+
+  private SafeFuture<List<DepositContract.DepositEventEventResponse>> mockContractEventsInRange(
+      long fromBlockNumber, long toBlockNumber) {
+    SafeFuture<List<DepositContract.DepositEventEventResponse>> safeFuture = new SafeFuture<>();
+    doReturn(safeFuture)
+        .when(depositContract)
+        .depositEventEventsInRange(
+            argThat(
+                argument ->
+                    Numeric.decodeQuantity(argument.getValue())
+                        .equals(BigInteger.valueOf(fromBlockNumber))),
+            argThat(
+                argument ->
+                    Numeric.decodeQuantity(argument.getValue())
+                        .equals(BigInteger.valueOf(toBlockNumber))));
```

Comment:
nit: I think this can be written with the more usual `when(depositContract.depositEventEventsInRange(...)).thenReturn(...)` style which will preserve type safety.

Reply:
Done.

Reply:
Actually, I think Mockito has a problem. When I write it that way, it gives a null pointer exception.